src/streamlink/plugins/trovo.py

```diff
@@ -0,0 +1,167 @@
+"""
+$description Global video game live streaming and video hosting platform, owned by Tencent.
+$url trovo.live
+$type live, vod
+"""
+
+import logging
+import random
+import re
+
+from streamlink.plugin import Plugin, pluginmatcher
+from streamlink.plugin.api import validate
+from streamlink.stream.hls import HLSStream
+
+log = logging.getLogger(__name__)
+
+
+@pluginmatcher(re.compile(
+    r"https?://(?:www\.)?trovo\.live/([^/?&]+)(?:/([^/?&]+))?"
+))
+class Trovo(Plugin):
+    @staticmethod
+    def generate_qid():
+        return ("%010x" % random.getrandbits(40)).upper()
+
+    def get_vod(self, video_id):
+        json = self.session.http.post(
+            f"https://gql.trovo.live/?qid={self.generate_qid()}",
+            json=[{
+                "operationName": "batchGetVodDetailInfo",
+                "variables": {
+                    "params": {
+                        "vids": [video_id],
+                    },
+                },
+                "extensions": {
+                    "persistedQuery": {
+                        "version": 1,
+                        "sha256Hash": "ceae0355d66476e21a1dd8e8af9f68de95b4019da2cda8b177c9a2255dad31d0",
```

Comment:
Ah, I meant to mention that, but forgot.  These hashes seem to be computed from queries defined within the sourced JS.  I looked at trying to compute it and it looked rather difficult, not to mention it would be unmaintainable if it were done.  It does not appear to be an sha256 sum alone, the code loops through an array of integers in the sourced JS, amongst other things.  All rather dreadful.

The hash used here was obtained from the network traffic and has remained constant in the week or so that I have observed it.

Reply:
GraphQL persistedQuery IDs are just cached GQL queries on the server side. I didn't find this exact ID in my network activity while browsing some VODs, so the query could've already been changed and a different ID is now used.
https://www.apollographql.com/docs/apollo-server/performance/apq/
This isn't much of a problem now, but it could be if they change their database models. Let's just keep it for now.

Reply:
I looked at some of the docs on Apollo regarding the persisted queries last week.  In what I saw, they explain how it works, but not in great detail.  The implementation seems to be simple if you pull in their JS to take care of it all for you behind the scenes.  I did wonder if it would be possible to use the full base query, but I'm not familiar with GQL and wasn't sure about how to use what I found (if it was even the right thing).

Reply:
It's still there for me, by the way - in the payload data of the POST requests:

https://gql.trovo.live/?chunk=1&reqms=1650397043989&qid=9BF05FD913&cli=4&client_info={"device_info":"{\"tid\":\"16503970247684763\"}"}&from=/clip/lc-387702298160635191?ltab=videos&locale=en-US&resolution=760*1280
```
[
  {
    "operationName": "getVersionUpdateTips",
    "variables": {},
    "extensions": {
      "persistedQuery": {
        "version": 1,
        "sha256Hash": "669c9bb4d507500703acf252ee398ea17eb94f2d0b42b3d283c451e3aeb9bb8f"
      }
    }
  },
  {
    "operationName": "batchGetVodDetailInfo",
    "variables": {
      "params": {
        "vids": [
          "lc-387702298160635191"
        ]
      }
    },
    "extensions": {
      "persistedQuery": {
        "version": 1,
        "sha256Hash": "ceae0355d66476e21a1dd8e8af9f68de95b4019da2cda8b177c9a2255dad31d0"
      }
    }
  },
  ....
```
and:

https://gql.trovo.live/?chunk=1&reqms=1650397624783&qid=5C418F880F&cli=4&client_info={"device_info":"{\"tid\":\"16503976105681262\"}"}&from=/video/ltv-106639746_106639746_387702298489518804?ltab=videos&locale=en-US&resolution=760*1280
```
[
  {
    "operationName": "getVersionUpdateTips",
    "variables": {},
    "extensions": {
      "persistedQuery": {
        "version": 1,
        "sha256Hash": "669c9bb4d507500703acf252ee398ea17eb94f2d0b42b3d283c451e3aeb9bb8f"
      }
    }
  },
  {
    "operationName": "batchGetVodDetailInfo",
    "variables": {
      "params": {
        "vids": [
          "ltv-106639746_106639746_387702298489518804"
        ]
      }
    },
    "extensions": {
      "persistedQuery": {
        "version": 1,
        "sha256Hash": "ceae0355d66476e21a1dd8e8af9f68de95b4019da2cda8b177c9a2255dad31d0"
      }
    }
  },
  ....
```